src/persist/src/runtime.rs

```diff
@@ -125,97 +124,162 @@ where
             }
         }
     });
-
-    // Construct the client.
-    let handles = Mutex::new(Some(RuntimeHandles {
+    let handles = RuntimeHandle(RuntimeHandleInner::Full {
+        id,
         impl_handle,
         ticker_handle,
         ticker_pool: pool.clone(),
-    }));
-    let core = RuntimeCore {
-        handles,
-        tx,
-        metrics,
+    });
+
+    // Explictly drop the pool guard so the lifetime is obvious.
+    drop(pool_guard);
+
+    Ok(RuntimeClient::new(handles, tx, metrics))
+}
+
+/// Starts a read-only runtime in a [std::thread].
+///
+/// This returns a clone-able client handle. The runtime is stopped when any
+/// client calls [RuntimeClient::stop] or when all clients have been dropped.
+///
+/// If Some, the given [tokio::runtime::Runtime] is used for IO and cpu heavy
+/// operations. If None, a new Runtime is constructed for this. The latter
+/// requires that we are not in the context of an existing Runtime, so if this
+/// is the case, the caller must use the Some form.
+//
+// TODO: The rust doc above is still a bit of a lie. Actually use this runtime
+// for all IO and cpu heavy operations.
+//
+// TODO: The whole story around Runtime usage in persist is pretty awkward and
+// still pretty unprincipled. I think when we do the TODO to make the Log and
+// Blob storage traits async, this will clear up a bit.
+//
+// NB: This is pretty duplicative of start, but that's okay. It's already fairly
+// different and it's only going to get more so. DRYing this up wouldn't really
+// decrease the chance of bugs but it would decrease readability.
+pub fn start_read<L, B>(
```

Comment:
sorry if this is a stupid question, but: would this actually work as it is now? Could you create a read-only for a shared storage location from a separate process. This still takes a blob, which when created would clash with the lock of the "real" runtime? And this just has a regular blob/maintainer/indexed, which would theoretically be able to modify the original data.

I guess I'm just wondering about the state of this. Totally fine if the intend is to just sort out the client side. And I could be wrong above. :sweat_smile:

Reply:
oh yeah, I missed the `BlobRead` bound on this one. :see_no_evil:

Reply:
:thumbsup: